[JSC] Optimize string replace with cache

Constellation opened the pull request:
#### ac69818d7335cd9d9acb8bdf3432272f58e6b156
<pre>
[JSC] Optimize string replace with cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=289420">https://bugs.webkit.org/show_bug.cgi?id=289420</a>
<a href="https://rdar.apple.com/146584722">rdar://146584722</a>

Reviewed by Yijia Huang.

This patch optimizes string replace with cache by not having
sourceRanges. Since we have cached JSImmutableButterfly, we can
reconstruct the range without keeping them in a Vector. We also compute
totalLength and 8Bit status during iteration instead of doing it later
again.

* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::putByVal):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::replaceUsingRegExpSearchWithCache):

Canonical link: <a href="https://commits.webkit.org/291859@main">https://commits.webkit.org/291859@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6a146db83ff4c5604fd0ad0e0df1cee2bd089a7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94235 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13822 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99247 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/44762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14122 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22252 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/71886 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/44762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97237 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10479 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85082 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/52232 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10168 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/2767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44080 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/86945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80400 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/2857 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101291 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/92901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21287 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15503 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80890 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21539 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81095 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80272 "Passed tests") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/24821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2187 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15115 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21271 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26450 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115551 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20958 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24418 "Built successfully") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22699 "Failed to checkout and rebase branch from PR 42149") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -576,57 +570,147 @@ static ALWAYS_INLINE JSString* replaceUsingRegExpSearchWithCache(VM& vm, JSGloba
-            sourceRanges.constructAndAppend(lastIndex, start);
-            replacements[replacementIndex++] = WTFMove(string);
+                auto substring = sourceView.substring(lastIndex, start - lastIndex);
+                substring.getCharacters8(buffer.subspan(bufferPos));
```
lol, `getCharacters8` the naming is confusing. Maybe `copyCharacters8` is better.

Constellation replied:
Anyway, this is pre-existing code :)

hyjorc1 commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -576,57 +570,147 @@ static ALWAYS_INLINE JSString* replaceUsingRegExpSearchWithCache(VM& vm, JSGloba
+            }
 
-            sourceRanges.constructAndAppend(lastIndex, start);
+            RELEASE_AND_RETURN(scope, jsString(vm, impl.releaseNonNull()));
```
Maybe we can just add this before this line.
```
ASSERT(lastIndex <= sourceLen && bufferPos + (sourceLen - lastIndex) == totalLength);
```

hyjorc1 commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -576,57 +570,147 @@ static ALWAYS_INLINE JSString* replaceUsingRegExpSearchWithCache(VM& vm, JSGloba
+            substring.getCharacters(buffer.subspan(bufferPos));
+        }
+
+        RELEASE_AND_RETURN(scope, jsString(vm, impl.releaseNonNull()));
```
Maybe we can just add this before this line.
```
ASSERT(lastIndex <= sourceLen && bufferPos + (sourceLen - lastIndex) == totalLength);
```

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Optimize string replace with cache
https://bugs.webkit.org/show_bug.cgi?id=289420
rdar://146584722

Reviewed by Yijia Huang.

This patch optimizes string replace with cache by not having
sourceRanges. Since we have cached JSImmutableButterfly, we can
reconstruct the range without keeping them in a Vector. We also compute
totalLength and 8Bit status during iteration instead of doing it later
again.

* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::putByVal):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::replaceUsingRegExpSearchWithCache):

Canonical link: https://commits.webkit.org/291859@main
